pkg/server/plugin/keymanager/awskms/awskms.go

```diff
@@ -0,0 +1,533 @@
+	}
+
+	keyType, err := keyTypeFromKeySpec(describeResp.KeyMetadata.CustomerMasterKeySpec)
+	if err != nil {
+		p.log.Debug("Skipped key", "reason", err)
+		return nil, nil
+	}
+
+	getPublicKeyResp, err := p.kmsClient.GetPublicKey(ctx, &kms.GetPublicKeyInput{KeyId: aliasName})
+	if err != nil {
+		return nil, kmsErr.New("failed to get public key: %v", err)
+	}
+
+	return &keyEntry{
+		KMSKeyID: *awsKeyID,
+		Alias:    *aliasName,
+		PublicKey: &keymanager.PublicKey{
+			Id:       spireKeyID,
+			Type:     keyType,
+			PkixData: getPublicKeyResp.PublicKey,
+		},
+	}, err
+}
+
+// fetchAliasesPage gets all aliases from KMS, and builds an in memory representation, with the ones that belong to the current server
+func (p *Plugin) fetchAliasesPage(ctx context.Context, marker *string) (*string, error) {
+	aliasesResp, err := p.kmsClient.ListAliases(ctx, &kms.ListAliasesInput{
+		Marker: marker,
+	})
+	if err != nil {
+		return nil, kmsErr.New("failed to fetch keys: %v", err)
+	}
+
+	p.log.Debug("Found aliases", "num_aliases", len(aliasesResp.Aliases))
+
+	for _, alias := range aliasesResp.Aliases {
+		if alias.AliasName == nil {
+			continue
+		}
+
+		spireKeyID, ok := p.spireKeyIDFromAlias(*alias.AliasName)
+		if !ok {
+			continue
+		}
+
+		if alias.TargetKeyId == nil {
+			// this means something external to the plugin created the alias, without associating it to a key.
+			// it should never happen with CMKs.
+			return nil, kmsErr.New("found SPIRE alias without key: %q", *alias.AliasName)
+		}
+
+		entry, err := p.buildKeyEntry(ctx, alias.AliasName, alias.TargetKeyId, spireKeyID)
+		switch {
+		case err != nil:
+			return nil, kmsErr.New("failed to process KMS key: %v", err)
+		case entry != nil:
+			err := p.setEntry(entry.PublicKey.Id, *entry)
+			if err != nil {
+				return nil, err
+			}
+			p.log.Debug("Key loaded", keyIDTag, *alias.TargetKeyId, aliasTag, *alias.AliasName)
+		}
+	}
+	return aliasesResp.NextMarker, nil
+}
+
+func (p *Plugin) scheduleDeleteTask(ctx context.Context) {
+	t := p.clock.Ticker(30 * time.Second)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			p.log.Debug("Stopping schedule delete task", "reason", ctx.Err())
+			return
+		case <-t.C:
+			keyID, ok := <-p.scheduleDelete
+			if !ok {
+				continue
+			}
+			_, err := p.kmsClient.ScheduleKeyDeletion(ctx, &kms.ScheduleKeyDeletionInput{
+				KeyId:               aws.String(keyID),
+				PendingWindowInDays: aws.Int32(7),
+			})
+			if err != nil {
+				p.log.Error("It was not possible to schedule deletion for key", "reason", err, keyIDTag, keyID)
+				p.scheduleDelete <- keyID
+			}
+			p.log.Debug("Key deleted", keyIDTag, keyID)
+		}
+	}
+}
+
+func (p *Plugin) spireKeyIDFromAlias(alias string) (string, bool) {
+	prefix := aliasPrefix + p.keyPrefix
+	if !strings.HasPrefix(alias, prefix) {
+		return "", false
+	}
+	return strings.TrimPrefix(alias, prefix), true
+}
+
+func (p *Plugin) aliasFromSpireKeyID(spireKeyID string) string {
```

Comment:
nit: This function doesn't use any state from `p`, I'd suggest it be a standalone function. When I see struct methods I expect that they need additional state other than the input to do their job.

Reply:
woops you are right, it uses the prefix.